### PR TITLE
fix partial audio transcription order:

### PR DESCRIPTION
**Pull request by nithinraok**
# What does this PR do ?

* Removes special case of sorting for EncDecMultiTaskModels
* Removes partial audio transcription for all Model Types
* Now transcribe_speech.py transcribes as per presence of offset and duration in manifest automatically without special conditions
* Adds support to transcriptionmixin to read manifest for transcribing on high level rather than subclasses to deal with it
* Merges NeMo 2.0 transcription usage for all Model Types

**Collection**: ASR

# Usage

```bash
python transcribe_speech.py model_path=<.nemo file> dataset_manifest=<manifest_file>
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [x] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Comment by pzelasko** [comment]
wouldn't it be better to enable pre-sorting for partial audio instead (if possible)?

**Comment by nithinraok** [comment]
Thanks for the suggestion @pzelasko, that needs a whole revamp and done that. But now I notice we can unify partial audio. 
1. Canary can perform partial audio transcription when offset provided through manifest (previous assumption was no I believe looking at the code)
2. If 1, then we can add additional support for .transcribe() for ASR Models to support reading from manifest. Currently TranscriptioMixin supports audio, list of audio, dataloader.

File: examples/asr/transcribe_speech.py [hunk 1]
```diff
@@ -293,7 +284,7 @@ def main(cfg: TranscriptionConfig) -> Union[TranscriptionConfig, List[Hypothesis
     elif isinstance(asr_model, EncDecHybridRNNTCTCModel):
         if cfg.decoder_type and cfg.decoder_type not in ['ctc', 'rnnt']:
             raise ValueError('Hybrid model only support ctc or rnnt decoding!')
-    else:  # rnnt model, there could be other models needs to be addressed.
```
Comment (titu1994) [inline examples/asr/transcribe_speech.py:296]: Add a else case in case there's a future model type

File: examples/asr/transcribe_speech.py [hunk 2]
```diff
@@ -361,29 +352,11 @@ def main(cfg: TranscriptionConfig) -> Union[TranscriptionConfig, List[Hypothesis
...
-                        print(json.dumps(item), file=f)
-                    cfg.dataset_manifest = f.name
-                    remove_path_after_done = f.name
-            filepaths = cfg.dataset_manifest
-    else:
-        # prepare audio filepaths and decide wether it's partial audio
-        filepaths, partial_audio = prepare_audio_data(cfg)
+    filepaths, sorted_manifest_path = prepare_audio_data(cfg)
+
+    remove_path_after_done = sorted_manifest_path if sorted_manifest_path is not None else None
```
Comment (titu1994) [inline examples/asr/transcribe_speech.py:357]: What does this flag calculate ?

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 3]
```diff
@@ -783,17 +783,6 @@ def _transcribe_on_begin(self, audio, trcfg: MultiTaskTranscriptionConfig):
                     trcfg._internal.primary_language = self.tokenizer.langs[0]
                     logging.debug(f"Transcribing with default setting of {trcfg._internal.primary_language}.")
 
-        elif isinstance(audio, str):
```
Comment (titu1994) [inline nemo/collections/asr/models/aed_multitask_models.py:786]: Why are we removing manifest ability from multi task model ?

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 4]
```diff
@@ -957,9 +932,14 @@ def _setup_transcribe_dataloader(self, config: Dict) -> 'torch.utils.data.DataLo
         Returns:
             A pytorch DataLoader for the given audio file(s).
         """
-        batch_size = min(config['batch_size'], len(config['paths2audio_files']))
+        if 'manifest_filepath' in config:
+            manifest_filepath = config['manifest_filepath']
+            batch_size = config['batch_size']
+        else:
+            manifest_filepath = os.path.join(config['temp_dir'], 'manifest.json')
```
Comment (titu1994) [inline nemo/collections/asr/models/aed_multitask_models.py:939]: When is this else case occuring? Can you add a comment ?

File: nemo/collections/asr/parts/mixins/transcription.py [hunk 5]
```diff
@@ -480,6 +481,12 @@ def _transcribe_input_processing(self, audio, trcfg: TranscribeConfig):
 
         # Check if audio is a list of strings (filepaths or manifests)
         if isinstance(audio[0], str):
+            trcfg._internal.manifest_filepath = None
```
Comment (titu1994) [inline nemo/collections/asr/parts/mixins/transcription.py:484]: You need to add this argument to the internal config of transcribe above

**Review by titu1994 — COMMENTED** [review]
Requires minor changes, but overall nice cleanup of partial transcription

File: examples/asr/transcribe_speech.py (hunk 2 above)
Comment (nithinraok) [reply examples/asr/transcribe_speech.py:357]: This was added previously to clean the temporarily created manifest later.

File: nemo/collections/asr/models/aed_multitask_models.py (hunk 3 above)
Comment (nithinraok) [reply nemo/collections/asr/models/aed_multitask_models.py:786]: its not removing but same check is now already part of TranscriptionMixin to handle manifest

File: nemo/collections/asr/models/aed_multitask_models.py (hunk 4 above)
Comment (nithinraok) [reply nemo/collections/asr/models/aed_multitask_models.py:939]: This is from the Transcription mixin where temp_dir was created to create a temp manifest when audio files are passed through list to .transcribe()

File: nemo/collections/asr/parts/mixins/transcription.py (hunk 5 above)
Comment (nithinraok) [reply nemo/collections/asr/parts/mixins/transcription.py]: Done, thank you!

**Review by stevehuang52 — APPROVED** [review]
Looks good on my end, thanks~!
